tests/unittest/noctx.cpp

```diff
@@ -0,0 +1,484 @@
+void St30pHandler::rxSt30DefaultTimestampsCheck(int sessionIdx, void* frame,
+                                                size_t frame_size) {
+  st30_frame* f = (st30_frame*)frame;
+  static uint64_t last_timestamp = 0;
+  EXPECT_NEAR(st10_media_clk_to_ns(f->timestamp, 48000),
+              (sessionsUserData.idx_rx) * nsPacketTime, (nsPacketTime) / 5);
+
+  if (last_timestamp != 0) {
+    uint64_t diff = f->timestamp - last_timestamp;
+    EXPECT_TRUE(diff == st10_tai_to_media_clk(nsPacketTime, 48000));
+  }
+
+  last_timestamp = f->timestamp;
+  sessionsUserData.idx_rx++;
+}
+
+TEST_F(NoCtxTest, st30p_default_timestamps) {
+  int sessionIdx = 0;
+
+  /* This tests makes mtl start from 0 */
+  ctx->para.ptp_get_time_fn = NoCtxTest::TestPtpSourceSinceEpoch;
+  ctx->para.log_level = MTL_LOG_LEVEL_INFO;
+
+  ASSERT_TRUE(ctx && ctx->handle == nullptr);
+  ctx->handle = mtl_init(&ctx->para);
+  ASSERT_TRUE(ctx->handle != nullptr);
+
+  st30pHandlers.emplace_back(std::make_unique<St30pHandler>(sessionIdx, ctx));
+  auto& handler = st30pHandlers[sessionIdx];
+  handler->createSession();
+
+  handler->rxTestFrameModifier = [this, sessionIdx](int idx, void* frame,
+                                                    size_t frame_size) {
+    this->st30pHandlers[sessionIdx]->rxSt30DefaultTimestampsCheck(idx, frame, frame_size);
+  };
+
+  TestPtpSourceSinceEpoch(nullptr);
+
+
+  for (int i = 0; i < defaultTestDuration; ++i) {
+    if (HasFailure()) break;
+    sleep(1);
+  }
+
+  mtl_stop(ctx->handle);
+}
+
+void St30pHandler::rxSt30DefaultUserPacingCheck(int sessionIdx, void* frame,
+                                                size_t frame_size) {
+  st30_frame* f = (st30_frame*)frame;
+  static uint64_t startingTime = 10 * NS_PER_MS;
+  static uint64_t last_timestamp = 0;
+  sessionsUserData.idx_rx++;
+
+  uint64_t expectedTimestamp = startingTime + (nsPacketTime * (sessionsUserData.idx_rx - 1));
+  uint64_t expected_media_clk = st10_tai_to_media_clk(expectedTimestamp, 48000);
+
+  EXPECT_EQ(f->timestamp, expected_media_clk);
+
+  if (last_timestamp != 0) {
+    uint64_t diff = f->timestamp - last_timestamp;
+    EXPECT_TRUE(diff == st10_tai_to_media_clk(nsPacketTime, 48000));
+  }
+
+  last_timestamp = f->timestamp;
+}
+
+void St30pHandler::txSt30DefaultUserPacing(int sessionIdx, void* frame,
+                                           size_t frame_size) {
+  st30_frame* f = (st30_frame*)frame;
+  static uint64_t startingTime = 10 * NS_PER_MS;
+
+  f->tfmt = ST10_TIMESTAMP_FMT_TAI;
+  f->timestamp = startingTime + (nsPacketTime * (sessionsUserData.idx_tx));
+  sessionsUserData.idx_tx++;
+}
+
+TEST_F(NoCtxTest, st30p_user_pacing) {
+  int sessionIdx = 0;
+
+  /* This tests makes mtl start from 0 */
+  ctx->para.ptp_get_time_fn = NoCtxTest::TestPtpSourceSinceEpoch;
+  ctx->para.log_level = MTL_LOG_LEVEL_INFO;
+
+  ASSERT_TRUE(ctx && ctx->handle == nullptr);
+  ctx->handle = mtl_init(&ctx->para);
+  ASSERT_TRUE(ctx->handle != nullptr);
```

Comment:
you can hide this behind some method in NoCtxTest that will take instance parameters as input.

Reply:
or not. leaving option to create negative tests is also important.